snippets/zoe/contracts/test-loan.js

```diff
@@ -0,0 +1,141 @@
+
+import '@agoric/install-ses';
+import { makeFakeVatAdmin } from '@agoric/zoe/src/contractFacet/fakeVatAdmin';
+import { makeZoe } from '@agoric/zoe';
+import bundleSource from '@agoric/bundle-source';
+import { makeIssuerKit } from '@agoric/ertp';
+import test from 'ava';
+import { E } from '@agoric/eventual-send';
+import buildManualTimer from '@agoric/zoe/tools/manualTimer';
+import { makeFakePriceAuthority } from '@agoric/zoe/tools/fakePriceAuthority';
+import { makeSubscriptionKit } from '@agoric/notifier';
+
+test('loan contract', async t => {
+  const zoe = makeZoe(makeFakeVatAdmin().admin);
+
+  const contractBundle = await bundleSource(
+    require.resolve('@agoric/zoe/src/contracts/loan'),
+  );
+  const installation = await E(zoe).install(contractBundle);
+
+  const collateralKit = makeIssuerKit('moola');
+  const loanKit = makeIssuerKit('simoleans');
+
+  // Create autoswap installation and instance
+  const autoswapBundle = await bundleSource(
+    require.resolve('@agoric/zoe/src/contracts/autoswap'),
+  );
+  const autoswapInstallation = await E(zoe).install(autoswapBundle);
+
+  const { instance: autoswapInstance } = await E(zoe).startInstance(
+    autoswapInstallation,
+    harden({ Central: collateralKit.issuer, Secondary: loanKit.issuer }),
+  );
+
+  const issuerKeywordRecord = harden({
+    Collateral: collateralKit.issuer,
+    Loan: loanKit.issuer,
+  });
+
+  const timer = buildManualTimer(console.log);
+
+  const priceAuthority = makeFakePriceAuthority({
+    mathIn: collateralKit.amountMath,
+    mathOut: loanKit.amountMath,
+    priceList: [4, 2],
+    timer,
+  });
+
+  const doAddCollateral = _ => {};
+  const allCollateralAmount = collateralKit.amountMath.make(1000);
+  const x = loanKit.amountMath.make(1500);
+
+  // #region customMarginCall
+  E(priceAuthority)
+    .quoteWhenLT(allCollateralAmount, x)
+    .then(priceQuote => doAddCollateral(priceQuote));
+  // #endregion customMarginCall
+
+  const { subscription: periodAsyncIterable } = makeSubscriptionKit();
+
+  // #region lend
+  const terms = {
+    mmr: 150,
+    autoswapInstance,
+    priceAuthority,
+    periodAsyncIterable,
+    interestRate: 5,
+  };
+
+  const { creatorInvitation: lendInvitation } = await E(zoe).startInstance(
+    installation,
+    issuerKeywordRecord,
+    terms,
+  );
+
+  const maxLoan = loanKit.amountMath.make(1000);
+
+  const proposal = harden({
+    give: { Loan: maxLoan },
+  });
+
+  const payments = harden({
+    Loan: loanKit.mint.mintPayment(maxLoan),
+  });
+
+  const lenderSeat = await E(zoe).offer(lendInvitation, proposal, payments);
+
+  const borrowInvitation = await E(lenderSeat).getOfferResult();
+  // #endregion lend
+
+  // #region borrow
+  const borrowerProposal = harden({
+    want: { Loan: maxLoan },
+    give: { Collateral: allCollateralAmount },
+  });
+
+  const borrowerPayments = {
+    Collateral: collateralKit.mint.mintPayment(allCollateralAmount),
+  };
+  const borrowSeat = await E(zoe).offer(
```

Comment:
Does this need an `await`?

Reply:
Hmm, I could not `await`, but then I would have to explain that it is actually a promise for a seat.

Reply:
I think that's a good thing. The alternative is that readers have to notice that sometimes there's an await, and figure out why.